[ActionSheet] Add support dynamic type using UIFont scaled fonts

codeman7 opened the pull request:
## Context
Adds support for UIFont's preferredFont... APIs to render Dynamic Type
fonts. Adds conformance to UIContentSizeCategoryAdjusting so that the fonts
can be adjusted automatically if desired by clients.

## Testing
1. Open up MDCDragons app
2. Open the ActionSheet -> "Material : UIKit comparison" example
3. Change the system font size within settings application

Additionally snapshot test have been added.

Closes #8622

codeman7 committed:
pull from #8596

codeman7 committed:
Messing with the tableView.frame

Updated the example to be workable. Also, see if the tableView being the wrong frame fixes the bug. Currently it doesn't but this is something I need to look into in the future

codeman7 committed:
Remove title and message from example for debugging

The table view wasn't updating it's contentSize so added a call to  will need to clean this up. Still having issue where table is jumping to max height when we go from small content to large content but large content to small content is resizing correctly

material-automation commented:
bazel detected changes to the following targets:

```
//components/ActionSheet:snapshot_tests
//components/ActionSheet:snapshot_tests_test_bundle
//components/ActionSheet:unit_tests
//components/ActionSheet:unit_tests_IPAD_PRO_12_9_IN_9_3
//components/ActionSheet:unit_tests_IPAD_PRO_12_9_IN_9_3_test_bundle
//components/ActionSheet:unit_tests_IPHONE_7_PLUS_IN_10_3
//components/ActionSheet:unit_tests_IPHONE_7_PLUS_IN_10_3_test_bundle
//components/ActionSheet:unit_tests_IPHONE_X_IN_11_0
//components/ActionSheet:unit_tests_IPHONE_X_IN_11_0_test_bundle
```

<!-- identifier: affected-targets -->

material-automation commented on components/ActionSheet/examples/ActionSheetComparisonExampleViewController.m:
```diff
@@ -82,8 +82,8 @@ - (void)viewDidLayoutSubviews {
-      [MDCActionSheetController actionSheetControllerWithTitle:@"Action Sheet Title"
-                                                       message:@"Action Sheet Message"];
+      [MDCActionSheetController actionSheetControllerWithTitle:nil
+                                                       message:nil];
```
clang-format suggested the following change:


```
-      [MDCActionSheetController actionSheetControllerWithTitle:nil
-                                                       message:nil];
+      [MDCActionSheetController actionSheetControllerWithTitle:nil message:nil];
```

<!-- identifier: clang-format -->
<!-- File: components/ActionSheet/examples/ActionSheetComparisonExampleViewController.m -->

material-automation commented on components/ActionSheet/examples/ActionSheetComparisonExampleViewController.m:
```diff
@@ -96,16 +96,48 @@ - (void)showMaterialActionSheet {
+  [actionSheet addAction:emailAction4];
   [actionSheet applyThemeWithScheme:self.containerScheme];
+//  actionSheet.title = @"Choose Your Adventure";
+//  actionSheet.message = @"Consider carefully before you make your choice. Your next step may lead to riches or utter defeat!";
```
clang-format suggested the following change:


```
-//  actionSheet.title = @"Choose Your Adventure";
-//  actionSheet.message = @"Consider carefully before you make your choice. Your next step may lead to riches or utter defeat!";
+  //  actionSheet.title = @"Choose Your Adventure";
+  //  actionSheet.message = @"Consider carefully before you make your choice. Your next step may
+  //  lead to riches or utter defeat!";
```

<!-- identifier: clang-format -->
<!-- File: components/ActionSheet/examples/ActionSheetComparisonExampleViewController.m -->

material-automation commented on components/ActionSheet/examples/ActionSheetComparisonExampleViewController.m:
```diff
@@ -96,16 +96,48 @@ - (void)showMaterialActionSheet {
+  if (@available(iOS 11.0, *)) {
+    actionSheet.actionFont = [[UIFontMetrics metricsForTextStyle:UIFontTextStyleBody] scaledFontForFont:font];
+//    actionSheet.titleFont = [[UIFontMetrics metricsForTextStyle:UIFontTextStyleTitle2] scaledFontForFont:font];
+//    actionSheet.messageFont = [[UIFontMetrics metricsForTextStyle:UIFontTextStyleSubheadline] scaledFontForFont:font];
```
clang-format suggested the following change:


```
-    actionSheet.actionFont = [[UIFontMetrics metricsForTextStyle:UIFontTextStyleBody] scaledFontForFont:font];
-//    actionSheet.titleFont = [[UIFontMetrics metricsForTextStyle:UIFontTextStyleTitle2] scaledFontForFont:font];
-//    actionSheet.messageFont = [[UIFontMetrics metricsForTextStyle:UIFontTextStyleSubheadline] scaledFontForFont:font];
+    actionSheet.actionFont =
+        [[UIFontMetrics metricsForTextStyle:UIFontTextStyleBody] scaledFontForFont:font];
+    //    actionSheet.titleFont = [[UIFontMetrics metricsForTextStyle:UIFontTextStyleTitle2]
+    //    scaledFontForFont:font]; actionSheet.messageFont = [[UIFontMetrics
+    //    metricsForTextStyle:UIFontTextStyleSubheadline] scaledFontForFont:font];
```

<!-- identifier: clang-format -->
<!-- File: components/ActionSheet/examples/ActionSheetComparisonExampleViewController.m -->

material-automation commented on components/ActionSheet/src/MDCActionSheetController.m:
```diff
@@ -180,14 +183,24 @@ - (void)viewDidLoad {
   CGSize size = [self.header sizeThatFits:CGRectStandardize(self.view.bounds).size];
+  CGFloat totalHeight = self.tableView.contentSize.height + size.height + dividerHeight;
+  if (@available(iOS 10.0, *)) {
+    NSLog(@"Table view's contentSize = %f\nTraitCollection.contentSize = %@", self.tableView.contentSize.height, self.traitCollection.preferredContentSizeCategory);
```
clang-format suggested the following change:


```
-    NSLog(@"Table view's contentSize = %f\nTraitCollection.contentSize = %@", self.tableView.contentSize.height, self.traitCollection.preferredContentSizeCategory);
+    NSLog(@"Table view's contentSize = %f\nTraitCollection.contentSize = %@",
+          self.tableView.contentSize.height, self.traitCollection.preferredContentSizeCategory);
```

<!-- identifier: clang-format -->
<!-- File: components/ActionSheet/src/MDCActionSheetController.m -->

material-automation commented on components/ActionSheet/src/MDCActionSheetController.m:
```diff
@@ -180,14 +183,24 @@ - (void)viewDidLoad {
+    NSLog(@"Table view's contentSize = %f\nTraitCollection.contentSize = %@", self.tableView.contentSize.height, self.traitCollection.preferredContentSizeCategory);
+  }
+  if (totalHeight > (CGRectGetHeight(self.view.bounds) / 2)) {
+    self.mdc_bottomSheetPresentationController.preferredSheetHeight = CGRectGetHeight(self.view.bounds) / 2;
```
clang-format suggested the following change:


```
-    self.mdc_bottomSheetPresentationController.preferredSheetHeight = CGRectGetHeight(self.view.bounds) / 2;
+    self.mdc_bottomSheetPresentationController.preferredSheetHeight =
+        CGRectGetHeight(self.view.bounds) / 2;
```

<!-- identifier: clang-format -->
<!-- File: components/ActionSheet/src/MDCActionSheetController.m -->

material-automation commented:
clang-format recommended changes.

Consider installing [git-clang-format](https://github.com/llvm-mirror/clang/blob/master/tools/clang-format/git-clang-format) and running the following command:

```bash
git clang-format $(git merge-base origin/develop HEAD)
```

<!-- identifier: clang-format -->

codeman7 committed:
Add the title and message back

I think the header isn't reporting back the correct , other than that the original PR should be good

material-automation commented on components/ActionSheet/examples/ActionSheetComparisonExampleViewController.m:
```diff
@@ -96,16 +96,48 @@ - (void)showMaterialActionSheet {
+  [actionSheet addAction:emailAction4];
   [actionSheet applyThemeWithScheme:self.containerScheme];
+  actionSheet.title = @"Choose Your Adventure";
+  actionSheet.message = @"Consider carefully before you make your choice. Your next step may lead to riches or utter defeat!";
```
clang-format suggested the following change:


```
-  actionSheet.message = @"Consider carefully before you make your choice. Your next step may lead to riches or utter defeat!";
+  actionSheet.message = @"Consider carefully before you make your choice. Your next step may lead "
+                        @"to riches or utter defeat!";
```

<!-- identifier: clang-format -->
<!-- File: components/ActionSheet/examples/ActionSheetComparisonExampleViewController.m -->

material-automation commented on components/ActionSheet/examples/ActionSheetComparisonExampleViewController.m:
```diff
@@ -96,16 +96,48 @@ - (void)showMaterialActionSheet {
+  if (@available(iOS 11.0, *)) {
+    actionSheet.actionFont = [[UIFontMetrics metricsForTextStyle:UIFontTextStyleBody] scaledFontForFont:font];
+    actionSheet.titleFont = [[UIFontMetrics metricsForTextStyle:UIFontTextStyleTitle2] scaledFontForFont:font];
+    actionSheet.messageFont = [[UIFontMetrics metricsForTextStyle:UIFontTextStyleSubheadline] scaledFontForFont:font];
```
clang-format suggested the following change:


```
-    actionSheet.actionFont = [[UIFontMetrics metricsForTextStyle:UIFontTextStyleBody] scaledFontForFont:font];
-    actionSheet.titleFont = [[UIFontMetrics metricsForTextStyle:UIFontTextStyleTitle2] scaledFontForFont:font];
-    actionSheet.messageFont = [[UIFontMetrics metricsForTextStyle:UIFontTextStyleSubheadline] scaledFontForFont:font];
+    actionSheet.actionFont =
+        [[UIFontMetrics metricsForTextStyle:UIFontTextStyleBody] scaledFontForFont:font];
+    actionSheet.titleFont =
+        [[UIFontMetrics metricsForTextStyle:UIFontTextStyleTitle2] scaledFontForFont:font];
+    actionSheet.messageFont =
+        [[UIFontMetrics metricsForTextStyle:UIFontTextStyleSubheadline] scaledFontForFont:font];
```

<!-- identifier: clang-format -->
<!-- File: components/ActionSheet/examples/ActionSheetComparisonExampleViewController.m -->

material-automation commented on components/ActionSheet/src/MDCActionSheetController.m:
```diff
@@ -180,14 +183,26 @@ - (void)viewDidLoad {
   CGSize size = [self.header sizeThatFits:CGRectStandardize(self.view.bounds).size];
+  CGFloat totalHeight = self.tableView.contentSize.height + size.height + dividerHeight;
+  if (@available(iOS 10.0, *)) {
+    NSLog(@"Table view's contentSize = %f\nHeader's size = %f\nTraitCollection.contentSize = %@", self.tableView.contentSize.height, size.height, self.traitCollection.preferredContentSizeCategory);
```
clang-format suggested the following change:


```
-    NSLog(@"Table view's contentSize = %f\nHeader's size = %f\nTraitCollection.contentSize = %@", self.tableView.contentSize.height, size.height, self.traitCollection.preferredContentSizeCategory);
+    NSLog(@"Table view's contentSize = %f\nHeader's size = %f\nTraitCollection.contentSize = %@",
+          self.tableView.contentSize.height, size.height,
+          self.traitCollection.preferredContentSizeCategory);
```

<!-- identifier: clang-format -->
<!-- File: components/ActionSheet/src/MDCActionSheetController.m -->

codeman7 committed:
Fix bug

Fix issue where the header wasn't resizing correctly. This bug was caused by the font not being set in . By changing the layout code to be in  the font is set so then  works as expected. Also remove the log statements.

codeman7 committed:
revert example changes

codeman7 committed:
remove extra property

codeman7 committed:
Remove unneeded changes

codeman7 committed:
clang

codeman7 committed:
:camera:

codeman7 committed:
add tests

codeman7 committed:
:camera:

codeman7 committed:
snapshots

codeman7 committed:
Update snapshots

romoore commented on components/ActionSheet/examples/ActionSheetComparisonExampleViewController.m:
```diff
@@ -106,6 +106,17 @@ - (void)showMaterialActionSheet {
   [actionSheet addAction:emailAction];
   [actionSheet applyThemeWithScheme:self.containerScheme];
+
+  UIFont *font = [UIFont fontWithName:@"Zapfino" size:8];
```
Were these left in by mistake? I don't believe the examples should demonstrate using Zapfino and we can just stick with the current examples.  Snapshot tests should be sufficient to verify the UIFontMetrics support.

codeman7 replied:
I left these here to showcase the new behavior but didn't feel strongly about it. They have been removed.

romoore commented on components/ActionSheet/src/MDCActionSheetController.m:
```diff
@@ -177,17 +179,18 @@ - (void)viewDidLoad {
 
-- (void)viewWillLayoutSubviews {
-  [super viewWillLayoutSubviews];
+- (void)viewDidLayoutSubviews {
```
Please split this change into a separate PR. It looks like it fixes a bug detectable with the current snapshot tests, so it should be pretty easy to split out and submit.  Then the PR for UIFontMetrics can be submitted in isolation.  In case either of these changes causes regressions internally, we won't lose the other if we  roll-back.

codeman7 replied:
These were broken out into #8735

romoore commented on components/ActionSheet/src/private/MDCActionSheetHeaderView.h:
```diff
@@ -36,4 +36,10 @@
 @property(nonatomic, strong, nullable) UIColor *messageTextColor;
 
+/** Shows the title of the Action Sheet. */
+@property(nonatomic, nonnull, strong, readonly) UILabel *titleLabel;
```
If you're making this change, please remove the `font` and `color` properties and just assign them directly to the labels' properties.

codeman7 replied:
I've updated the header to conform to the protocol and reverted these changes. Please take another look here.

romoore replied:
Ack.

romoore commented on components/ActionSheet/src/private/MDCActionSheetItemTableViewCell.h:
```diff
@@ -31,6 +31,9 @@
 @property(nonatomic, strong, nullable) UIColor *actionTextColor;
 
+/** Label for the action's title text. */
+@property(nonatomic, nonnull, readonly, strong) UILabel *actionLabel;
```
Same request here - now that the label is exposed, please remove the properties that passed-through to the label.

codeman7 replied:
I've updated the cell to conform to the protocol and reverted these changes. Please take another look here.

romoore replied:
Ack.

romoore commented on components/ActionSheet/tests/snapshot/MDCActionSheetControllerSnapshotTests.m:
```diff
@@ -968,4 +984,117 @@ - (void)testActionSheetWithHeaderShownButNoTitleOrMessage {
+    // Then
+    // Can't add a UIWindow to a UIView, so just screenshot the window directly.
+    [window layoutIfNeeded];
+    [self snapshotVerifyView:window];
```
Tip: I would suggest snapshotting the actionsheet's view instead. In the past when I attempted to snapshot a Window (for Dialogs) it caused a crash in a different test.

codeman7 replied:
Done

romoore reviewed (changes requested):
I think the internal fixes should be made separately from the Dynamic Type changes. That will help isolate any breaks caused by either change when the release lands.  If you think they need to land together, let's chat offline about it.

codeman7 committed:
Merge branch 'develop' into action-sheet-dt

codeman7 committed:
Address feedback / Refactor

Instead of moving the subviews to the .h file have the header conform to the protocol

codeman7 committed:
clean up example changes

material-automation commented on components/ActionSheet/src/private/MDCActionSheetHeaderView.h:
```diff
@@ -14,7 +14,7 @@
 #import <UIKit/UIKit.h>
 
-@interface MDCActionSheetHeaderView : UIView
+@interface MDCActionSheetHeaderView : UIView<UIContentSizeCategoryAdjusting>
```
clang-format suggested the following change:


```
-@interface MDCActionSheetHeaderView : UIView<UIContentSizeCategoryAdjusting>
+@interface MDCActionSheetHeaderView : UIView <UIContentSizeCategoryAdjusting>
```

<!-- identifier: clang-format -->
<!-- File: components/ActionSheet/src/private/MDCActionSheetHeaderView.h -->

material-automation commented:
clang-format recommended changes.

Consider installing [git-clang-format](https://github.com/llvm-mirror/clang/blob/master/tools/clang-format/git-clang-format) and running the following command:

```bash
git clang-format $(git merge-base origin/develop HEAD)
```

<!-- identifier: clang-format -->

codeman7 committed:
Change the cell

I still need to clean this up but going to take a break

codeman7 committed:
Update cell to match header

material-automation commented on components/ActionSheet/src/private/MDCActionSheetItemTableViewCell.h:
```diff
@@ -17,7 +17,7 @@
 #import "MaterialInk.h"
 
-@interface MDCActionSheetItemTableViewCell : UITableViewCell
+@interface MDCActionSheetItemTableViewCell : UITableViewCell<UIContentSizeCategoryAdjusting>
```
clang-format suggested the following change:


```
-@interface MDCActionSheetItemTableViewCell : UITableViewCell<UIContentSizeCategoryAdjusting>
+@interface MDCActionSheetItemTableViewCell : UITableViewCell <UIContentSizeCategoryAdjusting>
```

<!-- identifier: clang-format -->
<!-- File: components/ActionSheet/src/private/MDCActionSheetItemTableViewCell.h -->

codeman7 committed:
clean up the snapshot tests

codeman7 committed:
:tool:

romoore commented on components/ActionSheet/src/MDCActionSheetController.m:
```diff
@@ -423,6 +438,14 @@ - (void)setShowsHeaderDivider:(BOOL)showsHeaderDivider {
+  if (@available(iOS 10.0, *)) {
+    _adjustsFontForContentSizeCategory = adjustsFontForContentSizeCategory;
+    self.header.adjustsFontForContentSizeCategory = _adjustsFontForContentSizeCategory;
+    [self updateTable];
```
Tip: Does the change in fonts require a full reload of data, or is it sufficient to just call `setNeedsLayout` on  `tableView`?

romoore commented on components/ActionSheet/src/private/MDCActionSheetItemTableViewCell.m:
```diff
@@ -138,6 +138,9 @@ - (MDCActionSheetAction *)action {
 }
 
 - (void)setActionFont:(UIFont *)actionFont {
+  if ([actionFont isEqual:_actionFont]) {
```
This looks like it might break support for `MDCFontLoader` fonts since it doesn't check if `mdc_scalingCurve` is equivalent (or set). 

Is there a way to avoid reassigning fonts when the trait collection changes that doesn't require this check in `setActionFont:`?

romoore commented on components/ActionSheet/src/private/MDCActionSheetItemTableViewCell.m:
```diff
@@ -155,7 +158,18 @@ - (void)updateTitleFont {
+}
+
 - (void)mdc_setAdjustsFontForContentSizeCategory:(BOOL)adjusts {
+  if (_mdc_adjustsFontForContentSizeCategory == adjusts) {
```
Is this required for the UIContentSizeCategoryAdjusting behavior or can it be split to another PR?

romoore commented on components/ActionSheet/tests/snapshot/MDCActionSheetControllerSnapshotTests.m:
```diff
@@ -968,4 +984,105 @@ - (void)testActionSheetWithHeaderShownButNoTitleOrMessage {
+    UIFont *bodyFont = [bodyMetrics scaledFontForFont:originalFont
+                        compatibleWithTraitCollection:extraSmallTraits];
+    self.actionSheetController.actionFont = bodyFont;
+    self.actionSheetController.view.bounds = CGRectMake(0, 0, 1000, 1000);
```
Tip: Is there any reason why these snapshots need to be 1000x1000? Can you not use the preferredContentSize or something similar to size the view?

romoore reviewed (changes requested):
I'm worried some of the optimizations around font setting might break MDCFontLoader-based fonts. Perhaps we can chat offline tomorrow about the need for the equality check?

codeman7 commented:
Not currently working on this might come back to it at a later date.
